src/herder/HerderSCPDriver.cpp

```diff
@@ -711,22 +723,82 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
                 {
                     highestTxSet = cTxSet;
                     highest = sv.txSetHash;
+                    if (candidateCloseTimes.size() == 0)
+                    {
+                        throw std::logic_error(
+                            "Missing candidate closeTimes list");
+                    }
+                    highestTxSetIndex =
+                        make_optional<size_t>(candidateCloseTimes.size() - 1);
+                    if (candidateCloseTimes[*highestTxSetIndex] != sv.closeTime)
+                    {
+                        throw std::logic_error("Incorrect highestTxSet index "
+                                               "into candidate closeTimes");
+                    }
+                    if (svSigned)
+                    {
+                        if (sv.ext.v() != STELLAR_VALUE_SIGNED)
+                        {
+                            throw std::logic_error(
+                                "Unsigned candidate StellarValue");
+                        }
+                        highestTxSetSignature =
+                            make_optional<LedgerCloseValueSignature>(
+                                sv.ext.lcValueSignature());
+                    }
                 }
             }
         }
+
+        if (!highestTxSet)
+        {
+            throw std::logic_error("No highest txSet");
+        }
         // make a copy as we're about to modify it and we don't want to mess
         // with the txSet cache
         bestTxSet = std::make_shared<TxSetFrame>(*highestTxSet);
     }
 
+    std::vector<TransactionFrameBasePtr> removed;
+    if (!highestTxSetIndex)
+    {
+        throw std::logic_error("Failed to record best transaction set index");
+    }
+    comp.closeTime = combinedCloseTime(candidateCloseTimes, *highestTxSetIndex);
+    if (comp.closeTime <= lcl.header.scpValue.closeTime)
+    {
+        throw std::logic_error(
+            "Tried to combine candidate with too-early close time");
+    }
+    if (highestTxSetSignature)
+    {
+        comp.ext.v(STELLAR_VALUE_SIGNED);
+        comp.ext.lcValueSignature() = *highestTxSetSignature;
+    }
+    else
+    {
+        comp.ext.v(STELLAR_VALUE_BASIC);
+    }
+    // This ->trimInvalid() should be a no-op:  we should have trimmed any
+    // expired transactions before nominating a candidate, and should have
+    // validated that other nodes have done so.  We call it as an assertion.
+    removed = bestTxSet->trimInvalid(
```

Comment:
this code is confusing and the only thing that can come out of it is a crash: in protocol 14 we don't modify the value, so there is no need to verify that it's still valid (check itself can be wrong)

Reply:
Alright, I've removed it.  Since nothing ever sets `removed` now before checking its `size()`, I've also removed the whole  `if (size() != 0)` block below.